HMM, static analysis still failed :thinking: 

```
Run vendor/bin/psalm  --output-format=github --shepherd --stats
Scanning files...
Analyzing files...

░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  60 / 185 (32%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 120 / 185 (64%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 180 / 185 (97%)
░░░░░
Error: Syntax error, unexpected T_STRING, expecting T_VARIABLE on line 9 (see https://psalm.dev/173)
🐑 results sent to shepherd.dev 🐑
Error: Process completed with exit code 2.
```

--- a/psalm-baseline.xml
+++ b/psalm-baseline.xml
@@ -423,9 +423,8 @@
     <MixedArgumentTypeCoercion occurrences="1">
       <code>$name</code>
     </MixedArgumentTypeCoercion>
-    <MixedAssignment occurrences="2">
+    <MixedAssignment occurrences="1">
       <code>$defaultValue</code>
-      <code>$value</code>
     </MixedAssignment>
     <PossiblyFalseArgument occurrences="1">
       <code>$reflectionProperty-&gt;getDocBlock()</code>
@@ -1585,8 +1584,7 @@
     </PossiblyNullArgument>
   </file>
   <file src="test/Generator/PropertyGeneratorTest.php">
-    <InvalidArgument occurrences="2">
-      <code>[PropertyGenerator::FLAG_CONSTANT, $flag]</code>
+    <InvalidArgument occurrences="1">
       <code>new stdClass()</code>
     </InvalidArgument>
     <InvalidReturnStatement occurrences="1"/>
@@ -1596,6 +1594,9 @@
     <MissingReturnType occurrences="1">
       <code>testOmitType</code>
     </MissingReturnType>
+    <MixedAssignment occurrences="1">
+      <code>$generator</code>
+    </MixedAssignment>
     <MixedInferredReturnType occurrences="1">
       <code>Generator</code>
     </MixedInferredReturnType>
@@ -1686,11 +1687,9 @@
     </UnusedVariable>
   </file>
   <file src="test/Generator/TestAsset/TestClassWithManyProperties.php">
-    <MissingPropertyType occurrences="6">
+    <MissingPropertyType occurrences="4">
       <code>$_barProperty</code>
       <code>$_barStaticProperty</code>
-      <code>$_bazProperty</code>
-      <code>$_bazStaticProperty</code>
       <code>$_complexType</code>
       <code>$fooStaticProperty</code>
     </MissingPropertyType>

--- a/src/Generator/AbstractMemberGenerator.php
+++ b/src/Generator/AbstractMemberGenerator.php
@@ -157,7 +157,7 @@ abstract class AbstractMemberGenerator extends AbstractGenerator
     }
 
     /**
-     * @return string
+     * @psalm-return static::VISIBILITY_*
      */
     public function getVisibility()
     {

--- a/src/Generator/PropertyGenerator.php
+++ b/src/Generator/PropertyGenerator.php
@@ -4,6 +4,12 @@ namespace Laminas\Code\Generator;
 
 use Laminas\Code\Reflection\PropertyReflection;
 
+use function array_reduce;
+use function get_class;
+use function gettype;
+use function is_bool;
+use function is_object;
+use function method_exists;
 use function sprintf;
 use function str_replace;
 use function strtolower;
@@ -11,6 +17,7 @@ use function strtolower;
 class PropertyGenerator extends AbstractMemberGenerator
 {
     public const FLAG_CONSTANT = 0x08;
+    public const FLAG_READONLY = 0x80;
 
     protected bool $isConst = false;
 
@@ -18,9 +25,7 @@ class PropertyGenerator extends AbstractMemberGenerator
 
     private bool $omitDefaultValue = false;
 
-    /**
-     * @return static
-     */
+    /** @return static */
     public static function fromReflection(PropertyReflection $reflectionProperty)
     {
         $property = new static();
@@ -41,6 +46,10 @@ class PropertyGenerator extends AbstractMemberGenerator
 
         if ($reflectionProperty->isStatic()) {
             $property->setStatic(true);
+        }
+
+        if (method_exists($reflectionProperty, 'isReadonly') && $reflectionProperty->isReadonly()) {
+            $property->setReadonly(true);
         }
 
         if ($reflectionProperty->isPrivate()) {
@@ -68,6 +77,7 @@ class PropertyGenerator extends AbstractMemberGenerator
      * @configkey static             bool
      * @configkey visibility         string
      * @configkey omitdefaultvalue   bool
+     * @configkey readonly           bool
      * @throws Exception\InvalidArgumentException
      * @param  array $array
      * @return static
@@ -112,6 +122,20 @@ class PropertyGenerator extends AbstractMemberGenerator
                 case 'omitdefaultvalue':
                     $property->omitDefaultValue($value);
                     break;
+                case 'readonly':
+                    if (! is_bool($value)) {
+                        throw new Exception\InvalidArgumentException(sprintf(
+                            '%s is expecting boolean on key %s. Got %s',
+                            __METHOD__,
+                            $name,
+                            is_object($value)
+                                ? get_class($value)
+                                : gettype($value)
+                        ));
+                    }
+
+                    $property->setReadonly($value);
+                    break;
             }
         }
 
@@ -119,12 +143,13 @@ class PropertyGenerator extends AbstractMemberGenerator
     }
 
     /**
-     * @param ?string $name
      * @param PropertyValueGenerator|string|array|null $defaultValue
-     * @param int $flags
+     * @param int|int[] $flags
      */
-    public function __construct($name = null, $defaultValue = null, $flags = self::FLAG_PUBLIC)
+    public function __construct(?string $name = null, $defaultValue = null, $flags = self::FLAG_PUBLIC)
     {
+        parent::__construct();
+
         if (null !== $name) {
             $this->setName($name);
         }
@@ -142,12 +167,12 @@ class PropertyGenerator extends AbstractMemberGenerator
      */
     public function setConst($const)
     {
-        if ($const) {
+        if (true === $const) {
             $this->setFlags(self::FLAG_CONSTANT);
-        } else {
-            $this->removeFlag(self::FLAG_CONSTANT);
+            return $this;
         }
 
+        $this->removeFlag(self::FLAG_CONSTANT);
         return $this;
     }
 
@@ -159,11 +184,47 @@ class PropertyGenerator extends AbstractMemberGenerator
         return (bool) ($this->flags & self::FLAG_CONSTANT);
     }
 
+    public function setReadonly(bool $readonly): self
+    {
+        if (true === $readonly) {
+            $this->setFlags(self::FLAG_READONLY);
+            return $this;
+        }
+
+        $this->removeFlag(self::FLAG_READONLY);
+        return $this;
+    }
+
+    public function isReadonly(): bool
+    {
+        return (bool) ($this->flags & self::FLAG_READONLY);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function setFlags($flags)
+    {
+        $flags = array_reduce((array) $flags, static function (int $a, int $b): int {
+            return $a | $b;
+        }, 0);
+
+        if ($flags & self::FLAG_READONLY && $flags & self::FLAG_STATIC) {
+            throw new Exception\RuntimeException('Modifier "readonly" in combination with "static" not permitted.');
+        }
+
+        if ($flags & self::FLAG_READONLY && $flags & self::FLAG_CONSTANT) {
+            throw new Exception\RuntimeException('Modifier "readonly" in combination with "constant" not permitted.');
+        }
+
+        return parent::setFlags($flags);
+    }
+
     /**
      * @param PropertyValueGenerator|mixed $defaultValue
      * @param string                       $defaultValueType
      * @param string                       $defaultValueOutputMode
-     * @return $this
+     * @return static
      */
     public function setDefaultValue(
         $defaultValue,
@@ -190,6 +251,7 @@ class PropertyGenerator extends AbstractMemberGenerator
     /**
      * @throws Exception\RuntimeException
      * @return string
+     * @psalm-return non-empty-string
      */
     public function generate()
     {
@@ -215,7 +277,11 @@ class PropertyGenerator extends AbstractMemberGenerator
                 . ($defaultValue !== null ? $defaultValue->generate() : 'null;');
         }
 
-        $output .= $this->indentation . $this->getVisibility() . ($this->isStatic() ? ' static' : '') . ' $' . $name;
+        $output .= $this->indentation
+            . $this->getVisibility()
+            . ($this->isReadonly() ? ' readonly' : '')
+            . ($this->isStatic() ? ' static' : '')
+            . ' $' . $name;
 
         if ($this->omitDefaultValue) {
             return $output . ';';

--- a/test/Generator/PropertyGeneratorTest.php
+++ b/test/Generator/PropertyGeneratorTest.php
@@ -12,12 +12,14 @@ use Laminas\Code\Generator\ValueGenerator;
 use Laminas\Code\Reflection\ClassReflection;
 use Laminas\Code\Reflection\PropertyReflection;
 use LaminasTest\Code\Generator\TestAsset\ClassWithTypedProperty;
+use PHP_CodeSniffer\Tokenizers\PHP;
 use PHPUnit\Framework\TestCase;
 use ReflectionProperty;
 use stdClass;
 
 use function array_shift;
 use function str_replace;
+use function uniqid;
 
 /**
  * @group Laminas_Code_Generator
@@ -141,9 +143,46 @@ EOS;
     }
 
     /**
+     * @dataProvider visibility
+     */
+    public function testPropertyCanProduceReadonlyModifier(int $flag, string $visibility): void
+    {
+        $codeGenProperty = new PropertyGenerator(
+            'someVal',
+            'some string value',
+            PropertyGenerator::FLAG_READONLY | $flag
+        );
+
+        self::assertSame(
+            '    ' . $visibility . ' readonly $someVal = \'some string value\';',
+            $codeGenProperty->generate()
+        );
+    }
+
+    public function testFailToProduceReadonlyStatic(): void
+    {
+        $codeGenProperty = new PropertyGenerator('someVal', 'some string value');
+
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('Modifier "readonly" in combination with "static" not permitted.');
+
+        $codeGenProperty->setFlags(PropertyGenerator::FLAG_READONLY | PropertyGenerator::FLAG_STATIC);
+    }
+
+    public function testFailToProduceReadonlyConstant(): void
+    {
+        $codeGenProperty = new PropertyGenerator('someVal', 'some string value');
+
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('Modifier "readonly" in combination with "constant" not permitted.');
+
+        $codeGenProperty->setFlags(PropertyGenerator::FLAG_READONLY | PropertyGenerator::FLAG_CONSTANT);
+    }
+
+    /**
      * @group PR-704
      */
-    public function testPropertyCanProduceContstantModifierWithSetter(): void
+    public function testPropertyCanProduceConstantModifierWithSetter(): void
     {
         $codeGenProperty = new PropertyGenerator('someVal', 'some string value');
         $codeGenProperty->setConst(true);
@@ -250,6 +289,7 @@ EOS;
 
         self::assertSame('SampleProperty', $propertyGenerator->getName());
         self::assertFalse($propertyGenerator->isConst());
+        self::assertFalse($propertyGenerator->isReadonly());
         self::assertInstanceOf(ValueGenerator::class, $propertyGenerator->getDefaultValue());
         self::assertInstanceOf(DocBlockGenerator::class, $propertyGenerator->getDocBlock());
         self::assertTrue($propertyGenerator->isAbstract());
@@ -263,6 +303,22 @@ EOS;
         $reflectionOmitDefaultValue->setAccessible(true);
 
         self::assertTrue($reflectionOmitDefaultValue->getValue($propertyGenerator));
+    }
+
+    public function testCreateReadonlyFromArray(): void
+    {
+        $propertyGenerator = PropertyGenerator::fromArray([
+            'name'     => 'ReadonlyProperty',
+            'readonly' => true,
+        ]);
+
+        self::assertSame('ReadonlyProperty', $propertyGenerator->getName());
+        self::assertFalse($propertyGenerator->isConst());
+        self::assertFalse($propertyGenerator->isAbstract());
+        self::assertFalse($propertyGenerator->isFinal());
+        self::assertFalse($propertyGenerator->isStatic());
+        self::assertTrue($propertyGenerator->isReadonly());
+        self::assertSame(PropertyGenerator::VISIBILITY_PUBLIC, $propertyGenerator->getVisibility());
     }
 
     /**
@@ -329,5 +385,20 @@ EOS;
         $code      = $generator->generate();
 
         self::assertSame('    private $typedProperty;', $code);
+    }
+
+    /** @requires PHP >= 8.1 */
+    public function testFromReflectionReadonlyProperty(): void
+    {
+        $className = uniqid('ClassWithReadonlyProperty', false);
+
+        eval('namespace ' . __NAMESPACE__ . '; class ' . $className . '{ public readonly string $readonly; }');
+
+        $reflectionProperty = new PropertyReflection(__NAMESPACE__ . '\\' . $className, 'readonly');
+
+        $generator = PropertyGenerator::fromReflection($reflectionProperty);
+        $code      = $generator->generate();
+
+        self::assertSame('    public readonly $readonly;', $code);
     }
 }

